staging/src/k8s.io/client-go/testing/fixture.go

```diff
@@ -398,37 +413,110 @@ func (t *tracker) add(gvr schema.GroupVersionResource, obj runtime.Object, ns st
 
 	if ns != newMeta.GetNamespace() {
 		msg := fmt.Sprintf("request namespace does not match object namespace, request: %q object: %q", ns, newMeta.GetNamespace())
-		return errors.NewBadRequest(msg)
+		return nil, errors.NewBadRequest(msg)
+	}
+	return obj, nil
+}
+
+func (t *tracker) create(gvr schema.GroupVersionResource, obj runtime.Object) error {
+	t.lock.Lock()
+	defer t.lock.Unlock()
+
+	objs, ok := t.objects[gvr]
+	if !ok {
+		objs = make(map[types.NamespacedName]runtime.Object)
+		t.objects[gvr] = objs
+	}
+
+	metaObj, err := meta.Accessor(obj)
+	if err != nil {
+		return err
+	}
+	namespacedName := types.NamespacedName{Namespace: metaObj.GetNamespace(), Name: metaObj.GetName()}
+	if _, ok = objs[namespacedName]; ok {
+		return errors.NewAlreadyExists(gvr.GroupResource(), metaObj.GetName())
+	}
+
+	objs[namespacedName] = obj
+
+	for _, w := range t.getWatches(gvr, metaObj.GetNamespace()) {
+		// To avoid the object from being accidentally modified by watcher
+		w.Add(obj.DeepCopyObject())
 	}
 
-	_, ok := t.objects[gvr]
+	return nil
+}
+
+func shouldDeleteDuringUpdate(oldObj, newObj runtime.Object) bool {
+	oldMeta, err := meta.Accessor(oldObj)
+	if err != nil {
+		panic(err)
+	}
+	newMeta, err := meta.Accessor(newObj)
+	if err != nil {
+		panic(err)
+	}
+	if len(newMeta.GetFinalizers()) > 0 {
+		// don't delete with finalizers remaining in the new object
+		return false
+	}
+	if oldMeta.GetDeletionTimestamp() == nil {
+		// don't delete if the existing object hasn't had a delete request made
+		return false
+	}
+	if ns, ok := newObj.(*v1.Namespace); ok && len(ns.Spec.Finalizers) > 0 {
+		// don't delete namespaces with finalizers remaining in the new object
+		return false
+	}
```

Comment:
I don't think we should special-case specific API types in the testing client like this

Reply:
I'm afraid we should preserve this. The unit test in namespace controller depends on this behavior (see the variable `testNamespacePendingFinalize`). I think this should be fine, because this is the only special case, and we should not add more in the future. This is also a special case in APIServer.